saq/queue/postgres.py

```diff
@@ -306,11 +305,42 @@ async def sweep(self, lock: int = 60, abort: float = 5.0) -> list[str]:
             cursor.nextset()
             cursor.nextset()
             results = await cursor.fetchall()
-            for result in results:
-                job = self.deserialize(result[0])
+            for key, job_bytes, objid in results:
+                job = self.deserialize(job_bytes)
                 if not job:
+                    swept.append(key)
+                    logger.info("Sweeping missing job %s", key)
+                    await cursor.execute(
+                        SQL(
+                            dedent(
+                                """
+                                DELETE FROM {jobs_table}
+                                WHERE key = %(key)s
+                                """
+                            )
+                        ).format(jobs_table=self.jobs_table),
+                        {"key": key},
+                    )
+                    continue
+                if job.status == Status.QUEUED:
+                    await asyncio.sleep(0.1)
```

Comment:
we can avoid this hack if there's a lock between dequeue and processing can we make that happen easily in postgres?

Reply:
I think the Postgres queue should already already handle it. I can probably remove the hack. When a job is dequeued, the job table's status column is updated to active in the same transaction by the queue, but the worker is the one that updates the status of the job object (separate transaction). The Postgres queue also grabs an advisory lock associated with the job though, and if the worker/queue's connection dies, the lock is released. The sweeper already takes care of active jobs without advisory locks.